proposals/p1885.md

```diff
@@ -0,0 +1,710 @@
+
+Mixins are currently in early design stages. This section highlights possible
+uses speculating on the final design. Some may include:
+
+-   Improved semantics for views compared to getter methods: no direct side
+    effect from using the view
+-   Facilitate code reuse, compared to reimplementing an interface
+-   Direct access to `self`, limiting needs for pointers and address resolution
+
+#### Mutable values
+
+Named mixins, used as programmable members, could expose a view with mutable
+values:
+
+```carbon
+// Using a named mixin to expose mutable values
+for (p: my_range.(Iterate.ElementType)* in my_range.mutable) {}
+```
+
+Used in this way, this clarifies that this a view of the data, like an
+attribute, with no direct side effects, when compared to a method.
+
+#### Alternative views
+
+Similarly, mixins could allow supporting other views into the container.
+
+For example, maps could expose a view with "key" and "value" only, in addition
+to the default (key, value) tuple.
+
+```carbon
+// Using a named mixin to expose mutable values
+for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
+```
+
+And "reversed" view could also be used for reverse iteration:
+
+```carbon
+// Reversing
+for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+```
+
+### Large elements and Optional
+
+Large elements would have a negative impact if the `Optional` cannot leverage
+unused bit patterns within the type, nor present a view instead of a copy.
+
+Such opportunities for `Optional` exist, and could be addressed in the future.
+
+### C++ interoperability
+
+The limitation of this approach is the need to adapt to C++ iterator model. This
+could be implemented by having the cursor be an iterator type:
+
+-   `NewCursor` providing the "start iterator"
+-   `Next` doing the "increment, bounds check, and returning value"
+
+Two different approaches are identified:
+
+-   One is to have a templated `impl` of `Iterate`, for anything with the right
+    method signatures ("begin", "end", etc.)
+-   Another is a templated adapter that implements `Iterate` when requested
+
+```carbon
+// template impl approach
+template constraint CppIterator { ... }
+template constraint CppContainer {
+  // speculative
+  forsome IteratorType: CppIterator;
+  fn begin() -> IteratorType;
+  fn end() -> IteratorType;
+}
+external impl forall [template T:! CppContainer] T as Iterate
+   where  .CursorType = T.(CppContainer.IteratorType)
+      and .ElementType = .CursorType.ElementType {
+  fn NewCursor[self: Self]() -> CursorType {
+    return self.(CppContainer.begin)();
+  }
+  fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType) {
+    if (*cursor == self.(CppContainer.end)()) {
+      return Optional(ElementType).MakeEmpty();
+    }
+    let value: ElementType = **cursor;
+    ++*cursor;
+    return Optional(ElementType).MakeValue(value);
+  }
+}
+```
+
+```carbon
+// adapter approach
+adapter CppIterate(template T:! type) for T {
```

Comment:
May be worth noting that this approach could have a cursor type parameter (with some sort of default computed from the return type of `T.begin()`), avoiding the deduction issues of the above approach.

Reply:
Added a mention to that effect.